src/legacy/ui/public/saved_objects/saved_object.ts

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * @name SavedObject
+ *
+ * NOTE: SavedObject seems to track a reference to an object in ES,
+ * and surface methods for CRUD functionality (save and delete). This seems
+ * similar to how Backbone Models work.
+ *
+ * This class seems to interface with ES primarily through the es Angular
+ * service and the saved object api.
+ */
+import { SavedObjectsClient, SavedObjectsClientContract } from 'kibana/public';
+import { npStart } from 'ui/new_platform';
+import { SavedObjectConfig, SavedObjectSaveOpts } from './types';
+import { buildSavedObject } from './helpers/build_saved_object';
+import { IndexPatternsContract } from '../../../../plugins/data/public';
+export { SavedObjectSaveOpts as SaveOptions } from './types';
+
+export interface SavedObject {
```

Comment:
I am slightly confused about the types here - is this `SavedObject` another one as the one in `types`? https://github.com/elastic/kibana/pull/51562/files#diff-00ac870029ae50508c1b15089d29cb16R23

It seems like it would be the same, and if its the case, then the `createSavedObjectClass` function could return the one from `types.ts` directly, then the re-definition like here: https://github.com/elastic/kibana/pull/51562/files#diff-e83ede3a0d730a36744c9420bf126a53R43 should not be necessary.

Typescript won't infer this because the ts-ignore in the `buildSavedObject` call in the constructor below, but you can cast it like this:
```ts
export function createSavedObjectClass(/*...*/) {
// ...
 class SavedObjectClass {
    constructor(config: SavedObjectConfig = {}) {
      buildSavedObject(
        // @ts-ignore
        this,
        config,
        indexPatterns,
        savedObjectsClient
      );
    }
  }

// this will now return a class that has all of the properties defined in the types file
  return SavedObjectClass as new (config: SavedObjectConfig) => SavedObject;
```

Reply:
good catch, the existing exported SavedObject type from this file was used by the plugin, I've cleaned this up, so there just one type less, and it's much less confusing. thx for this hint, really nice, you're Master of TypeScript 🥇 !